esphome/components/dxs238xw/dxs238xw.cpp

```diff
@@ -0,0 +1,1071 @@
+  this->load_initial_number_value_(this->preference_delay_value_set_, "delay_value_set",
+                                   &this->ms_data_.delay_value_set);
+  this->load_initial_number_value_(this->preference_energy_purchase_value_, "energy_purchase_value",
+                                   &this->lp_data_.energy_purchase_value_tmp);
+  this->load_initial_number_value_(this->preference_energy_purchase_alarm_, "energy_purchase_alarm",
+                                   &this->lp_data_.energy_purchase_alarm_tmp);
+  UPDATE_NUMBER(delay_value_set, this->ms_data_.delay_value_set)
+  UPDATE_NUMBER(energy_purchase_value, this->lp_data_.energy_purchase_value_tmp)
+  UPDATE_NUMBER(energy_purchase_alarm, this->lp_data_.energy_purchase_alarm_tmp)
+
+  ESP_LOGI(TAG, "* Load GET_METER_ID");
+  this->send_command_(SmCommandSend::GET_METER_ID);
+
+  ESP_LOGI(TAG, "* Load GET_POWER_STATE");
+  this->send_command_(SmCommandSend::GET_POWER_STATE);
+
+  ESP_LOGI(TAG, "* Load GET_LIMIT_AND_PURCHASE_DATA");
+  this->send_command_(SmCommandSend::GET_LIMIT_AND_PURCHASE_DATA);
+
+  ESP_LOGI(TAG, "* Load GET_MEASUREMENT_DATA");
+  this->send_command_(SmCommandSend::GET_MEASUREMENT_DATA);
+
+  this->status_momentary_warning("wait_to_loop_and_update", SM_WAIT_TO_LOOP_AND_UPDATE_STATE);
+}
+
+void Dxs238xwComponent::loop() {
+  if (this->get_component_state() == COMPONENT_STATE_LOOP) {
+    this->incoming_messages_();
+
+    this->send_command_(SmCommandSend::GET_POWER_STATE);
+    this->send_command_(SmCommandSend::GET_LIMIT_AND_PURCHASE_DATA);
+  }
+}
+
+void Dxs238xwComponent::update() {
+  if (this->get_component_state() == COMPONENT_STATE_LOOP) {
+    this->send_command_(SmCommandSend::GET_MEASUREMENT_DATA);
+  }
+}
+
+void Dxs238xwComponent::dump_config() {
+  this->check_uart_settings(SM_UART_CONFIG_BAUD_RATE, SM_UART_CONFIG_STOP_BITS, SM_UART_CONFIG_PARITY,
+                            SM_UART_CONFIG_DATA_BITS);
+
+  ESP_LOGCONFIG(TAG, "* energy_purchase_value: %u", this->lp_data_.energy_purchase_value_tmp);
+  ESP_LOGCONFIG(TAG, "* energy_purchase_Alarm: %u", this->lp_data_.energy_purchase_alarm_tmp);
+  ESP_LOGCONFIG(TAG, "* delay_value_set: %u", this->ms_data_.delay_value_set);
+}
+
+//------------------------------------------------------------------------------
+////////////////////////////////////////////////////////////////////////////////
+//------------------------------------------------------------------------------
+
+void Dxs238xwComponent::set_meter_state_toogle() {
+  bool state = !this->ms_data_.meter_state;
+
+  this->send_command_(SmCommandSend::SET_POWER_STATE, state);
+}
+
+void Dxs238xwComponent::set_meter_state_on() {
+  bool state = true;
+
+  this->send_command_(SmCommandSend::SET_POWER_STATE, state);
+}
+
+void Dxs238xwComponent::set_meter_state_off() {
+  bool state = false;
+
+  this->send_command_(SmCommandSend::SET_POWER_STATE, state);
+}
+
+void Dxs238xwComponent::send_hex_message(const char *message, bool check_crc) {
```

Comment:
Please use `std::string` here, and change the references

```suggestion
void Dxs238xwComponent::send_hex_message(std::string message, bool check_crc) {
```

Reply:
ok, i will change to std::string.